src/lib/storage/dictionary_segment.hpp

```
@@ -23,53 +25,93 @@ template <typename T>
+
+    // create dictionary
+    _dictionary = std::make_shared<std::vector<T>>(segment_values);
+    std::sort(_dictionary->begin(), _dictionary->end());
+    _dictionary->erase(std::unique(_dictionary->begin(), _dictionary->end()), _dictionary->end());
+
+    // create attribute vector with minimal width
+    if (unique_values_count() <= UINT8_MAX) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint8_t>>(segment->size());
+    } else if (unique_values_count() <= UINT16_MAX) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint16_t>>(segment->size());
+    } else {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint32_t>>(segment->size());
+    }
+
+    // fill attribute vector with valueIDs
+    for (ValueID position(0); position < segment->size(); position++) {
+      auto value_id = ValueID(std::distance(
+          _dictionary->begin(), std::find(_dictionary->begin(), _dictionary->end(), segment_values.at(position))));
+      _attribute_vector->set(position, value_id);
+    }
+  }
 
   // SEMINAR INFORMATION: Since most of these methods depend on the template parameter, you will have to implement
   // the DictionarySegment in this file. Replace the method signatures with actual implementations.
 
   // return the value at a certain position. If you want to write efficient operators, back off!
-  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override;
+  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override {
+    return _dictionary->at(_attribute_vector->get(chunk_offset));
+  }
 
   // return the value at a certain position.
-  T get(const size_t chunk_offset) const;
+  T get(const size_t chunk_offset) const { return value_by_value_id(ValueID(_attribute_vector->get(chunk_offset))); }
 
   // dictionary segments are immutable
-  void append(const AllTypeVariant&) override;
+  void append(const AllTypeVariant&) override {}
```

This will silently fail... not sure what our guidelines are, maybe rather throw an exception?